### PR TITLE
Blog post: Backwards Compatibility Testing for OpenSearch

### LINKED ISSUE #428: Backwards Compatibility Testing for OpenSearch
**Describe the blog post your would like to write**
A blog post for detailing the Backwards Compatibility testing framework in OpenSearch.

**What is the title of the blog post?**
Backwards Compatibility Testing for OpenSearch

**Who are the authors?**
@VachaShah 
@saratvemulapalli 

**What is the proposed posting date?**
11/8

### PR DESCRIPTION
**Pull request by VachaShah**
Signed-off-by: Vacha <vachshah@amazon.com>

### Description
Blog post for Backwards Compatibility Testing for OpenSearch.
 
### Issues Resolved
Closes #428 

By submitting this pull request, I confirm that my contribution is made under the terms of the BSD-3-Clause License.

### CODE REVIEW

File: _posts/2021-10-26-bwc-testing-for-opensearch.markdown [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+  - vachshah
+  - vemsarat
+date: 2021-10-26 01:01:01 -0700
+categories: 
+  - technical-post
+twittercard:
+  description: "This post provides details on the framework in OpenSearch for backwards compatibility testing"
+---
+
+Backwards Compatibility (BWC) testing is used to test and determine the safe upgrade paths from a supported BWC version to the current version. The framework in OpenSearch allows to run these BWC tests for all supported BWC versions in OpenSearch allowing safe upgrade paths between versions. This framework is now extended to work for plugins and they can introduce their BWC tests without creating individual frameworks of their own. This post provides details on the framework in OpenSearch for backwards compatibility.
```
Comment (stockholmux) [inline _posts/2021-10-26-bwc-testing-for-opensearch.markdown:14]: 'framework in OpenSearch allows **you** to run these BWC tests'
'work for plugins **which** can introduce their BWC tests without creating individual frameworks'

File: _posts/2021-10-26-bwc-testing-for-opensearch.markdown [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+## The Framework
+
+As a general idea for BWC tests, the framework supports to spin up a test cluster with a supported BWC version, then upgrade the nodes to the current version and test various features and functionalities as a result of the upgrade. This allows to test the compatibility of the code between versions.
```
Comment (stockholmux) [inline _posts/2021-10-26-bwc-testing-for-opensearch.markdown:18]: 'the framework supports **spinning** up a test cluster'
'This allows **for testing** the'

File: _posts/2021-10-26-bwc-testing-for-opensearch.markdown [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+## Test cluster setup
+
+The test clusters are available via  `opensearch.testclusters` gradle plugin. A cluster is an instance of `OpenSearchCluster` and it contains a list of nodes which are instances of the `OpenSearchNode` class which provides the methods to start, stop, restart, upgrade, install plugins etc. The version of the nodes can be set by providing a list of versions to the `testclusters`. The node can then be upgraded to a newer version provided by using the `nextNodeToNextVersion` method on the concerned testcluster or by using `goToNextVersion` method which upgrades all the nodes in a cluster.
+
+An example testcluster initialization is given below. Here the testcluster is named `mycluster` with 4 nodes. Two versions are provided and the first version is the version that gets installed by default when a node starts. It upgrades to the next version using the method mentioned above.
```
Comment (stockholmux) [inline _posts/2021-10-26-bwc-testing-for-opensearch.markdown:24]: back ticks around testcluster

File: _posts/2021-10-26-bwc-testing-for-opensearch.markdown [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    * Located in  `qa/mixed-cluster` 
+    * Run using  `./gradlew :qa:mixed-cluster:vM.N.b#mixedClusterTest`
+* **Rolling Upgrade:** For a rolling upgrade scenario, a test cluster is spun up with multiple nodes of a supported BWC version and each node is upgraded one by one to a newer version of OpenSearch using the `nextNodeToNextVersion` method resulting in a fulling upgraded cluster achieved via sequential upgrade. The corresponding tests are run for the old cluster and for each node after it undergoes the upgrade.
+    * Located in `qa/rolling-upgrade`
+    * Run using `./gradlew :qa:rolling-upgrade:vM.N.b#upgradedClusterTest`
+* **Full Cluster Restart:** For a full cluster restart, a test cluster is spun up with multiple nodes of a supported BWC version and all the nodes are upgraded using the [`goToNextVersion`](https://github.com/opensearch-project/OpenSearch/blob/main/buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchCluster.java#L388) method resulting in a fully upgrade cluster by restarting. The corresponding tests are run for the old cluster and for the upgraded cluster.
```
Comment (stockholmux) [inline _posts/2021-10-26-bwc-testing-for-opensearch.markdown:45]: 'fully **upgraded** cluster by restarting'

File: _posts/2021-10-26-bwc-testing-for-opensearch.markdown [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    * Located in `qa/full-cluster-restart`
+    * Run using `./gradlew :qa:full-cluster-restart:vM.N.b#upgradedClusterTest`
+* **Snapshot Upgrade**: For a snapshot upgrade, two test clusters are created (one with the older version and one with a newer version) and the snapshot restoration is tested between the clusters. Each step tests the creation of repository and snapshot and restoring the snapshot.
+    * Located in `qa/repository-multi-version`
+    * Run using `./gradlew :qa:repository-multi-version:vM.N.b#Step4NewClusterTest`
+
+To run all the backwards compatibility tests:
+ `./gradlew bwcTest`
```
Comment (stockholmux) [inline _posts/2021-10-26-bwc-testing-for-opensearch.markdown:53]: Add another line break after the code block - it runs together into one paragraph on the site.

File: _posts/2021-10-26-bwc-testing-for-opensearch.markdown [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+## Summary
+
+Backwards compatibility specially for plugins has been a major concern during releases since it involved manual testing to determine safe upgrade paths. With the extended framework for BWC, plugins can now implement their BWC tests thus alleviating the need for manual efforts. All pull requests in OpenSearch now run the BWC tests making sure all the changes adhere to the backwards compatibility standards.
```
Comment (stockholmux) [inline _posts/2021-10-26-bwc-testing-for-opensearch.markdown:87]: 'Backwards compatibility **especially** for plugins'

File: _posts/2021-10-26-bwc-testing-for-opensearch.markdown [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+                        return fileTree("`src/test/resources/org/opensearch/ad/bwc/anomaly-detection`/1.13.0.0/").getSingleFile()
+                    }
+                }
+            }
+        }))
+    }
+}
+```
+
+Mixed cluster, rolling upgrade and full restart upgrade scenarios are supported for the BWC tests in plugins and the nodes in the cluster can be upgraded using the methods on the testcluster:  [`upgradeNodeAndPluginToNextVersion`](https://github.com/opensearch-project/OpenSearch/blob/main/buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchCluster.java#L413) for a single node upgrade at a time and [`upgradeAllNodesAndPluginsToNextVersion`](https://github.com/opensearch-project/OpenSearch/blob/main/buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchCluster.java#L395) for upgrading all the nodes at the same time. The tests corresponding to each step can be configured to run similar to explained above according to each upgrade scenario.
```
Comment (stockholmux) [inline _posts/2021-10-26-bwc-testing-for-opensearch.markdown:83]: 'configured to run **similarly as** explained above'

**Review by stockholmux — CHANGES_REQUESTED** [review]
Really nice post @VachaShah and @saratvemulapalli. It clear and makes sense - just a few little points of grammar and disfluency.

**Comment by saratvemulapalli** [comment]
> Really nice post @VachaShah and @saratvemulapalli. It clear and makes sense - just a few little points of grammar and disfluency.

Just realized @VachaShah added me as an author. She did all the work so credits go to her.

**Review by stockholmux — APPROVED** [review]
LGTM

**Comment by stockholmux** [comment]
I'm satisfied with the post - as soon as @dblock and @saratvemulapalli give the nod, we can publish.

**Review by saratvemulapalli — APPROVED** [review]

**Comment by saratvemulapalli** [comment]
Looks good to me, ship it!!

**Comment by stockholmux** [comment]
@dblock Any objections?

File: _posts/2021-10-26-bwc-testing-for-opensearch.markdown [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+layout: post
+title:  "Backwards Compatibility Testing for OpenSearch"
+authors: 
+  - vachshah
+  - vemsarat
+date: 2021-10-26 01:01:01 -0700
+categories: 
+  - technical-post
+twittercard:
+  description: "This post provides details on the framework in OpenSearch for backwards compatibility testing"
```
Comment (dblock) [inline _posts/2021-10-26-bwc-testing-for-opensearch.markdown:11]: Missing period, it's a full sentence.

**Review by dblock — APPROVED** [review]
This is good! We can publish this or consider some potential updates before we do:

1. I think we can say more about which versions we test bcw with more explicitly in the intro. Maybe "For example, OpenSearch 1.1.0 was tested for backwards compatibility for versions ...".
2. We have some execution framework in openseach-build that runs bcw tests, but it's not perfect. You could expand on how and when we'd like to run bcw tests and link related issues other than opensearch-build#90.

File: _posts/2021-10-26-bwc-testing-for-opensearch.markdown (hunk 8 above)
Comment (stockholmux) [reply _posts/2021-10-26-bwc-testing-for-opensearch.markdown:11]: @VachaShah Can you address @dblock's points above before lunch today (2021-11-05)? I'd love to get this out the door today.

Comment (saratvemulapalli) [reply _posts/2021-10-26-bwc-testing-for-opensearch.markdown:11]: @stockholmux , @VachaShah is out today. 
I don't know if I have enough permissions to make changes on her fork. 

May be can publish this and then tweak it next week?

Comment (stockholmux) [reply _posts/2021-10-26-bwc-testing-for-opensearch.markdown:11]: @saratvemulapalli OK. Let's do that!

Comment (VachaShah) [reply _posts/2021-10-26-bwc-testing-for-opensearch.markdown:11]: Thank you @stockholmux @saratvemulapalli @dblock ! I will put up a PR for the tweak this week.

Comment (VachaShah) [reply _posts/2021-10-26-bwc-testing-for-opensearch.markdown:11]: @dblock I have created a PR #478 with the updates. Looks like opensearch-build#90 covers most of the bwc automation part, I added a little more along with it. LMK if I missed something.

**Pull request merged**